### PR TITLE
User update: Expanded the user patch API to persist the cadre details

### PR DESCRIPTION
**Pull request by ruksana2808**

### CODE REVIEW

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 1]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
 		return response;
 	}
 
+	private String validateCadreDetails(Map<String, Object> profileDetailsMap) {
+		List<Map<String, Object>> cadreDetails;
+
+		try {
+			if (profileDetailsMap.containsKey(Constants.CADRE_DETAILS)
+					&& profileDetailsMap.get(Constants.CADRE_DETAILS) != null) {
+				JsonNode cadreJson = mapper.valueToTree(profileDetailsMap.get(Constants.CADRE_DETAILS));
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: This is unnecessary... do not try to convert to Map to JsonNode -- literally both are similar but it is costlier.

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 2]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+				}
+				if (!cadreJson.has(Constants.CONTROLLING_AUTHORITY) || cadreJson.get(
+						Constants.CONTROLLING_AUTHORITY).isNull() || StringUtils.isEmpty(
+						cadreJson.get(Constants.CONTROLLING_AUTHORITY).asText())) {
+					return "Controlling authority of CadreDetails is missing";
+				}
+				if (!cadreJson.has(Constants.CADRE_BATCH) || cadreJson.get(Constants.CADRE_BATCH).isNull()
+						|| !cadreJson.get(Constants.CADRE_BATCH).isInt()) {
+					return "Cadre Batch of CadreDetails is missing and it should be an integer";
+				}
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: always try to return to most of the error as possible... 

Map<String, Object> cadreDetails = (Map<String, Object>) profileDetails.get(Constants.CADRE_DETAILS);
if (MapUtils.isEmpty(cadreDetails)) {
  return "Cadre Details object is empty";
}
List<String> mandatoryCaderAttributes = Array.asList("attrib1", attrib2");
List<String> missingAttribts = new ArrayList<String>();
for (String attrib : mandatoryCaderAttributes) {
   if (!cadreDetails.contains(attrib)) {
      missingAttribs.add(attrib);
   } else if (StringUtils.isBlank(cadreDetails.get(attrib)) {
     missingAttribs.add(attrib);
   }
 }
 
 if (missingAttrib.size() > 0) {
    return "Missing mandatory attributes. " + missingAttribs.toString());
 }

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 3]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+				Map<String, Object> propertyMap = new HashMap<>();
+				propertyMap.put(Constants.ID, Constants.CADRE_CONFIG);
+				cadreDetails = cassandraOperation.getRecordsByPropertiesWithoutFiltering(
+						Constants.KEYSPACE_SUNBIRD, Constants.TABLE_SYSTEM_SETTINGS, propertyMap,
+						null, 2);
+				if (cadreDetails.isEmpty() && !cadreDetails.contains(Constants.VALUE)) {
+					return "Details not found for this key";
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: error should be -> Failed to validate cadreDetails schema.

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 4]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+						cadreJson.get(Constants.CONTROLLING_AUTHORITY).asText())) {
+					return "Controlling authority of CadreDetails is missing";
+				}
+				if (!cadreJson.has(Constants.CADRE_BATCH) || cadreJson.get(Constants.CADRE_BATCH).isNull()
+						|| !cadreJson.get(Constants.CADRE_BATCH).isInt()) {
+					return "Cadre Batch of CadreDetails is missing and it should be an integer";
+				}
+				Map<String, Object> propertyMap = new HashMap<>();
+				propertyMap.put(Constants.ID, Constants.CADRE_CONFIG);
+				cadreDetails = cassandraOperation.getRecordsByPropertiesWithoutFiltering(
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: name this variable as "cadreSchemaDetails"

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 5]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+										if (cadreJson.get(Constants.CADRE_NAME).asText().equalsIgnoreCase(cadreName)) {
+											isCadrePresent = true;
+											break;
+										}
+									}
+								}
+							}
+
+						}
+					}
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: Need to optimize this block of code...

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 6]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
 		return response;
 	}
 
+	private String validateCadreDetails(Map<String, Object> profileDetailsMap) {
+		List<Map<String, Object>> cadreDetails;
+
+		try {
+			if (profileDetailsMap.containsKey(Constants.CADRE_DETAILS)
+					&& profileDetailsMap.get(Constants.CADRE_DETAILS) != null) {
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: we called this method bcz the given profileDetails had this cadreDetails in the requestBody... 
so simply get the object and check for null -- refer the comment for returning error messages...

**Review by karthik-tarento — APPROVED** [review]

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 7]
```diff
@@ -338,6 +351,108 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
 		return response;
 	}
 
+	private String validateCadreDetails(Map<String, Object> profileDetailsMap) {
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java:351]: Need to optimize this method

**Review by karthik-tarento — CHANGES_REQUESTED** [review]

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 1 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: changed

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 2 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: changed

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 5 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: changed

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 6 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: done

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 4 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: changed

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 3 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: done

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 8]
```diff
@@ -338,6 +350,150 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+		List<Map<String, Object>> cadreSchemaDetails;
+
+		try {
+			if (profileDetailsMap.get(Constants.CADRE_DETAILS) != null) {
+				Map<String, Object> cadreMap = mapper.convertValue(
+						profileDetailsMap.get(Constants.CADRE_DETAILS),
+						new TypeReference<Map<String, Object>>() {
+						});
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: no need to of ObjectMapper conversion required... simple typecast is enough...  
Map<String, Object> cadreMap = (Map<String, Object>) profileDetails.get(Constants.CADRE_DETAILS);

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java [hunk 9]
```diff
@@ -338,6 +350,150 @@ public SBApiResponse profileUpdate(Map<String, Object> request, String userToken
...
+					return "Invalid civilserviceType is given";
+				}
+				if (!isCivilTypePresent) {
+					return "Invalid civilSerivceName is given";
+				}
+				if (!isCadrePresent) {
+					return "Invalid cadreName is given";
+				}
+
+//				for (Map cadreDetail : cadreSchemaDetails) {
```
Comment (karthik-tarento) [inline src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: Remove the code instead of commenting out

**Review by karthik-tarento — CHANGES_REQUESTED** [review]

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 8 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: done

File: src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java (hunk 9 above)
Comment (ruksana2808) [reply src/main/java/org/sunbird/profile/service/ProfileServiceImpl.java]: done

**Pull request merged**